File: src/bytecode/compiler.ts
import type { Expression, Program } from "../ast.ts";
import { type Literal, TokenType } from "../token.ts";
import { err } from "../utils.ts";
import { Op } from "./bytecode.ts";

export interface Chunk {
  code: Uint8Array;
  constants: Literal[];
}

/**Compiler */
export class Compiler {
  private constants: Literal[] = [];
  private codes: number[] = [];
  private scopes: { vars: Map<string, number>; slotCount: number }[] = [{
    vars: new Map(),
    slotCount: 0,
  }];
  private nextSlot = 0;
  private funcs: Map<string, { addr: number; paramCount: number }>[] = [
    new Map(),
  ];
  private labels: Map<string, { high: number; low: number }> = new Map();

  emit = (op: Op, ...arg: number[]): void => {
    this.codes.push(op);
    if (arg !== undefined) this.codes.push(...arg);
  };

  private enterScope = () => {
    this.scopes.push({ vars: new Map(), slotCount: 0 });
    this.funcs.push(new Map());
  };

  private exitScope = () => {
    const scope = this.scopes.pop();
    if (scope) {
      this.nextSlot -= scope.slotCount;
    }
    this.funcs.pop();
  };

  private loadVar = (name: string): number | null => {
    for (let i = this.scopes.length - 1; i >= 0; i--) {
      const slot = this.scopes[i].vars.get(name);
      if (slot !== undefined) return slot;
    }
    return null;
  };

  private loadFunc = (
    name: string,
  ): { addr: number; paramCount: number } | null => {
    for (let i = this.funcs.length - 1; i >= 0; i--) {
      const func = this.funcs[i].get(name);
      if (func !== undefined) return func;
    }
    return null;
  };

  private declVar = (name: string): number => {
    const currentScope = this.scopes[this.scopes.length - 1];
    const slot = this.nextSlot++;
    currentScope.vars.set(name, slot);
    currentScope.slotCount++;
    return slot;
  };

  private modVar = (name: string): number => {
    const slot = this.loadVar(name);
    if (slot === null) {
      return err(
        "Compiler",
        `Variable '${name}' has not been defined`,
      );
    }
    return slot;
  };

  public compile = (program: Program): { chunk: Chunk; maxSlot: number } => {
    for (const expr of program.body) {
      this.compileExpr(expr);
    }

    this.emit(Op.HALT);

    return {
      chunk: {
        code: new Uint8Array(this.codes),
        constants: this.constants,
      },
      maxSlot: this.nextSlot,
    };
  };

  private compileExpr = (
    expr: Expression,
  ): void => {
    switch (expr.type) {
      case "Val": {
        const val = expr.value;
        this.constants.push(val);
        this.emit(Op.LOAD_CONST, this.constants.length - 1);
        break;
      }
      case "Var": {
        const slot = this.loadVar(expr.name);
        if (slot === null) {
          return err(
            "Compiler",
            `Variable '${expr.name}' has not been defined`,
          );
        }

        this.emit(Op.LOAD_VAR, slot);
        break;
      }
      case "VarDecl": {
        this.compileExpr(expr.value);
        const slot = this.declVar(expr.name);
        this.emit(Op.STORE_VAR, slot);
        this.emit(Op.POP);
        break;
      }
      case "VarMod": {
        this.compileExpr(expr.value);
        const slot = this.modVar(expr.name);
        this.emit(Op.STORE_VAR, slot);
        this.emit(Op.POP);
        break;
      }
      case "BinOp": {
        this.compileExpr(expr.left);
        this.compileExpr(expr.right);
        switch (expr.op) {
          case TokenType.OP_ADD: {
            this.emit(Op.ADD);
            break;
          }
          case TokenType.OP_SUB: {
            this.emit(Op.SUB);
            break;
          }
          case TokenType.OP_MUL:
            this.emit(Op.MUL);
            break;
          case TokenType.OP_DIV: {
            this.emit(Op.DIV);
            break;
          }
          case TokenType.COMP_EQ: {
            this.emit(Op.EQ);
            break;
          }
          case TokenType.COMP_NE: {
            this.emit(Op.NE);
            break;
          }
          case TokenType.COMP_GT: {
            this.emit(Op.GT);
            break;
          }
          case TokenType.COMP_GE: {
            this.emit(Op.GE);
            break;
          }
          case TokenType.COMP_LT: {
            this.emit(Op.LT);
            break;
          }
          case TokenType.COMP_LE: {
            this.emit(Op.LE);
            break;
          }
        }
        break;
      }
      case "UnaryOp": {
        this.compileExpr(expr.argument);
        if (expr.op === TokenType.OP_INC || expr.op === TokenType.OP_DEC) {
          if (expr.argument.type === "Var") {
            const name = expr.argument.name;
            const slot = this.loadVar(name)!;
            if (slot === undefined) {
              return err(
                "Compiler",
                `Variable '${name}' has not been defined`,
              );
            }

            this.emit(Op.LOAD_VAR, slot);
            if (expr.op === TokenType.OP_INC) {
              this.emit(Op.INC);
            } else {
              this.emit(Op.DEC);
            }
            this.emit(Op.STORE_VAR, slot);
            break;
          }
        }
        switch (expr.op) {
          case TokenType.LOG_NOT: {
            this.emit(Op.LOG_NOT);
            break;
          }
          default: {
            expr.op === TokenType.OP_NEG
              ? this.emit(Op.NEG)
              : this.emit(Op.POS);
          }
        }
        break;
      }
      case "Out": {
        this.compileExpr(expr.value);
        this.emit(Op.OUT);
        break;
      }
      case "In": {
        const slot = this.declVar(expr.name);
        this.emit(Op.IN, slot);
        break;
      }
      case "Stmt": {
        this.enterScope();
        const body = expr.body;

        for (let i = 0; i < body.length - 1; i++) {
          this.compileExpr(body[i]);
          this.emit(Op.POP);
        }

        if (body.length > 0) {
          this.compileExpr(body[body.length - 1]);
        } else {
          this.constants.push(undefined);
          this.emit(Op.LOAD_CONST, this.constants.length - 1);
        }

        this.exitScope();
        break;
      }
      case "If": {
        this.compileExpr(expr.cond);

        const thenPos = this.codes.length;
        this.emit(Op.JUMP_IF_FALSE, 0, 0);

        this.enterScope();
        this.compileExpr(expr.body);
        this.exitScope();

        let elsePos = -1;
        if (expr.else) {
          elsePos = this.codes.length;
          this.emit(Op.JUMP, 0, 0);
        }

        const thenEndPos = this.codes.length;
        this.patchJumpAddr(thenPos + 1, thenEndPos);

        if (expr.else) {
          this.enterScope();
          this.compileExpr(expr.else);
          this.exitScope();
          const elseEndPos = this.codes.length;
          this.patchJumpAddr(elsePos + 1, elseEndPos);
        }
        break;
      }

      case "While": {
        this.enterScope();

        const loopPos = this.codes.length;
        this.compileExpr(expr.cond);

        const jumpIfFalse = this.codes.length;
        this.emit(Op.JUMP_IF_FALSE, 0, 0);

        this.compileExpr(expr.body);
        this.emit(Op.JUMP, (loopPos >> 8) & 0xff, loopPos & 0xff);

        const breakPos = this.codes.length;
        this.patchJumpAddr(jumpIfFalse + 1, breakPos);

        this.exitScope();
        break;
      }
      case "FuncDecl": {
        const jumpAddr = this.codes.length;
        this.emit(Op.JUMP, 0, 0);
        const funcAddr = this.codes.length;
        const currFunc = this.funcs[this.funcs.length - 1];

        if (currFunc.has(expr.name)) {
          return err(
            "Compiler",
            `Function '${expr.name}' has already been declared in this scope`,
          );
        }

        currFunc.set(expr.name, {
          addr: funcAddr,
          paramCount: expr.params.length,
        });

        this.enterScope();

        for (const param of expr.params) {
          this.declVar(param);
        }

        this.compileExpr(expr.body);

        this.emit(Op.RET);

        this.exitScope();
        this.patchJumpAddr(jumpAddr + 1, this.codes.length);
        break;
      }
      case "FuncCall": {
        for (const arg of expr.args) {
          this.compileExpr(arg);
        }

        const func = this.loadFunc(expr.name);

        if (func === null) {
          return err(
            "Compiler",
            `Function '${expr.name}' hasn't been declared in this scope`,
          );
        }
        if (func.paramCount !== expr.args.length) {
          return err(
            "Compiler",
            `Function ${expr.name} expected ${func.paramCount} arguments, got ${expr.args.length}`,
          );
        }

        const target = func.addr;

        this.emit(
          Op.CALL,
          (target >> 8) & 0xff,
          target & 0xff,
          func.paramCount,
        );

        break;
      }
      case "Exit": {
        this.compileExpr(expr.status);
        this.emit(Op.EXIT);
        break;
      }
      case "Return": {
        if (expr.value) {
          this.compileExpr(expr.value);
        } else {
          this.constants.push(undefined);
          this.emit(Op.LOAD_CONST, this.constants.length - 1);
        }
        this.emit(Op.RET);
        break;
      }
      case "Eval": {
        this.compileExpr(expr.code);
        this.emit(Op.EVAL);
        break;
      }
      case "Label": {
        const addr = this.codes.length;
        this.labels.set(expr.name, {
          high: (addr >> 8) & 0xff,
          low: addr & 0xff,
        });
        break;
      }
      case "Goto": {
        const label = this.labels.get(expr.name);
        if (label === undefined) {
          return err("Compiler", `Label ${expr.name} not found`);
        }
        this.emit(Op.JUMP, label.high, label.low);
        break;
      }
      default: {
        return err(
          "Compiler",
          `Unknown node type: ${(expr as Expression).type}`,
        );
      }
    }
  };

  private patchJumpAddr = (pos: number, addr: number): void => {
    this.codes[pos] = (addr >> 8) & 0xff;
    this.codes[pos + 1] = addr & 0xff;
  };
}

File: src/bytecode/compiler.rs
use core::panic;
use std::collections::HashMap;

use serde::{Deserialize, Serialize};

use crate::{
    ast::{Expr, Program},
    bytecode::Op,
    token::{Literal, TokenType},
};

#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct Bytecode {
    pub chunk: Chunk,
    pub max_slot: u32,
}

#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct Chunk {
    pub code: Vec<u8>,
    pub constants: Vec<Literal>,
}

struct Scope {
    vars: HashMap<String, u32>,
    slot_count: u32,
}

struct Func {
    addr: u32,
    param_count: u32,
}

struct Label {
    high: u32,
    low: u32,
}

pub struct Compiler {
    constants: Vec<Literal>,
    code: Vec<u8>,
    scopes: Vec<Scope>,
    next_slot: u32,
    funcs: Vec<HashMap<String, Func>>,
    labels: HashMap<String, Label>,
}

impl Compiler {
    pub fn new() -> Self {
        Self {
            constants: Vec::new(),
            code: Vec::new(),
            scopes: Vec::new(),
            next_slot: 0,
            funcs: Vec::new(),
            labels: HashMap::new(),
        }
    }

    fn emit(&mut self, op: Op, args: &[u32]) -> () {
        self.code.push(op as u8);
        for arg in args {
            self.code.push(*arg as u8);
        }
    }

    fn enter_scope(&mut self) -> () {
        self.scopes.push(Scope {
            vars: HashMap::new(),
            slot_count: 0,
        });
        self.funcs.push(HashMap::new())
    }

    fn exit_scope(&mut self) -> () {
        if let Some(scope) = self.scopes.pop() {
            self.next_slot -= scope.slot_count;
        }
        self.funcs.pop();
    }

    fn load_var(&self, name: String) -> Option<&u32> {
        self.scopes
            .iter()
            .rev()
            .find_map(|scope| scope.vars.get(&name))
    }

    fn load_func(&self, name: String) -> Option<Func> {
        for func_map in self.funcs.iter().rev() {
            if let Some(func) = func_map.get(&name) {
                return Some(Func {
                    addr: func.addr,
                    param_count: func.param_count,
                });
            }
        }
        None
    }

    fn decl_var(&mut self, name: String) -> u32 {
        let curr_scope = self.scopes.last_mut().unwrap();
        let slot = self.next_slot;
        self.next_slot += 1;
        curr_scope.vars.insert(name, slot);
        curr_scope.slot_count += 1;
        slot
    }

    fn mod_var(&mut self, name: String) -> u32 {
        let slot = self.load_var(name.clone());
        match slot {
            Some(s) => *s,
            None => panic!("Compiler: Variable {} not found", name),
        }
    }

    pub fn compile(&mut self, program: Program) -> Bytecode {
        self.enter_scope();

        for expr in program.body {
            self.compile_expr(expr)
        }
        self.emit(Op::HALT, &[]);

        return Bytecode {
            chunk: Chunk {
                code: self.code.clone(),
                constants: self.constants.clone(),
            },
            max_slot: self.next_slot,
        };
    }

    fn compile_expr(&mut self, expr: Expr) -> () {
        match expr {
            Expr::Val(val) => {
                let value = val.value;
                self.constants.push(value);
                self.emit(Op::LOADCONST, &[(self.constants.len() - 1) as u32]);
            }
            Expr::Var(var) => {
                let slot = self.load_var(var.name.clone());
                match slot {
                    Some(s) => {
                        self.emit(Op::LOADVAR, &[*s]);
                    }
                    None => {
                        panic!("Compiler: Variable {} not found", var.name);
                    }
                }
            }
            Expr::VarDecl(decl) => {
                self.compile_expr(*decl.value);
                let slot = self.decl_var(decl.name);
                self.emit(Op::STOREVAR, &[slot]);
                self.emit(Op::POP, &[]);
            }
            Expr::VarMod(decl) => {
                self.compile_expr(*decl.value);
                let slot = self.mod_var(decl.name);
                self.emit(Op::STOREVAR, &[slot]);
                self.emit(Op::POP, &[]);
            }
            Expr::BinOp(bin) => {
                self.compile_expr(*bin.left);
                self.compile_expr(*bin.right);
                match bin.operator {
                    TokenType::ADD => self.emit(Op::ADD, &[]),
                    TokenType::SUB => self.emit(Op::SUB, &[]),
                    TokenType::MUL => self.emit(Op::MUL, &[]),
                    TokenType::DIV => self.emit(Op::DIV, &[]),
                    TokenType::LOGAND => self.emit(Op::LOGAND, &[]),
                    TokenType::LOGOR => self.emit(Op::LOGOR, &[]),
                    TokenType::LOGXOR => self.emit(Op::LOGXOR, &[]),
                    TokenType::COMPEQ => self.emit(Op::EQ, &[]),
                    TokenType::COMPNE => self.emit(Op::NE, &[]),
                    TokenType::COMPLT => self.emit(Op::LT, &[]),
                    TokenType::COMPGT => self.emit(Op::GT, &[]),
                    TokenType::COMPLE => self.emit(Op::LE, &[]),
                    TokenType::COMPGE => self.emit(Op::GE, &[]),
                    TokenType::COMPAND => self.emit(Op::AND, &[]),
                    TokenType::COMPOR => self.emit(Op::OR, &[]),
                    _ => {
                        panic!("Compiler: Unimplemented binary operator {:?}", bin.operator);
                    }
                }
            }
            Expr::UnaryOp(unary) => {
                self.compile_expr(*unary.argument.clone());
                match unary.operator {
                    TokenType::NEG => self.emit(Op::NEG, &[]),
                    TokenType::LOGNOT => self.emit(Op::LOGNOT, &[]),
                    TokenType::INC => match *unary.argument.clone() {
                        Expr::Var(var) => {
                            let name = var.name.clone();
                            let slot = self.mod_var(name.clone());
                            self.emit(Op::LOADVAR, &[slot]);
                            self.emit(Op::INC, &[]);
                            self.emit(Op::STOREVAR, &[slot]);
                        }
                        _ => panic!("Compiler: Unary INC operator only supports variables"),
                    },
                    TokenType::DEC => match *unary.argument.clone() {
                        Expr::Var(var) => {
                            let name = var.name.clone();
                            let slot = self.mod_var(name.clone());
                            self.emit(Op::LOADVAR, &[slot]);
                            self.emit(Op::DEC, &[]);
                            self.emit(Op::STOREVAR, &[slot]);
                        }
                        _ => panic!("Compiler: Unary DEC operator only supports variables"),
                    },
                    _ => {
                        panic!(
                            "Compiler: Unimplemented unary operator {:?}",
                            unary.operator
                        );
                    }
                }
            }
            Expr::Stmt(stmt) => {
                self.enter_scope();
                let body = stmt.body;

                for i in 0..body.len() - 1 {
                    self.compile_expr(body[i].clone());
                    self.emit(Op::POP, &[]);
                }

                if body.len() > 0 {
                    self.compile_expr(body.last().unwrap().clone());
                } else {
                    self.constants.push(Literal::Void);
                    self.emit(Op::LOADCONST, &[self.constants.len() as u32 - 1]);
                }

                self.exit_scope();
            }
            Expr::If(i) => {
                self.compile_expr(*i.condition);

                let then_branch_addr = self.code.len() as u32;
                self.emit(Op::JUMPIFFALSE, &[0, 0]);

                self.enter_scope();
                self.compile_expr(*i.then);
                self.exit_scope();

                let mut else_branch_addr: u32 = 1;
                if let Some(_) = i.else_branch.clone() {
                    else_branch_addr = self.code.len() as u32;
                    self.emit(Op::JUMP, &[0, 0]);
                }

                let then_end_addr = self.code.len() as u32;
                self.patch_jump_addr(then_branch_addr + 1, then_end_addr);

                if let Some(else_branch) = i.else_branch {
                    self.enter_scope();
                    self.compile_expr(*else_branch);
                    self.exit_scope();

                    let else_end_addr = self.code.len() as u32;
                    self.patch_jump_addr(else_branch_addr + 1, else_end_addr);
                }
            }
            Expr::While(w) => {
                self.enter_scope();

                let loop_pos = self.code.len() as u32;
                self.compile_expr(*w.condition.clone());

                let jump_if_false = self.code.len() as u32;
                self.emit(Op::JUMPIFFALSE, &[0, 0]);

                self.compile_expr(*w.body.clone());
                self.emit(
                    Op::JUMP,
                    &[((loop_pos >> 8) & 0xff) as u32, loop_pos & 0xFF],
                );

                let break_pos = self.code.len() as u32;
                self.patch_jump_addr(jump_if_false + 1, break_pos);

                self.exit_scope();
            }
            Expr::FuncDecl(decl) => {
                let jump_addr = self.code.len() as u32;
                self.emit(Op::JUMP, &[0, 0]);
                let func_addr = self.code.len() as u32;
                let curr_func = self.funcs.last_mut().unwrap();

                if curr_func.contains_key(&decl.name) {
                    panic!("Compiler: Function {} already declared", decl.name);
                }

                curr_func.insert(
                    decl.name.clone(),
                    Func {
                        addr: func_addr,
                        param_count: decl.params.len() as u32,
                    },
                );

                self.enter_scope();

                for param in decl.params {
                    self.decl_var(param);
                }

                self.compile_expr(*decl.body);
                self.emit(Op::RET, &[]);

                self.exit_scope();
                self.patch_jump_addr(jump_addr + 1, self.code.len() as u32);
            }
            Expr::FuncCall(call) => {
                for arg in call.args.clone() {
                    self.compile_expr(arg);
                }

                let func = self.load_func(call.name.clone());

                match func {
                    Some(f) => {
                        if f.param_count != call.args.len() as u32 {
                            panic!(
                                "Compiler: Function {} expects {} arguments, got {}",
                                call.name,
                                f.param_count,
                                call.args.len()
                            );
                        }

                        let target = f.addr;

                        self.emit(
                            Op::CALL,
                            &[
                                ((target >> 8) & 0xFF) as u32,
                                (target & 0xFF) as u32,
                                f.param_count,
                            ],
                        );
                    }
                    _ => {
                        panic!("Compiler: Function {} not found", call.name);
                    }
                }
            }
            Expr::Exit(exit) => {
                self.compile_expr(*exit.code);
                self.emit(Op::EXIT, &[]);
            }
            Expr::Return(ret) => {
                if let Some(value) = ret.value {
                    self.compile_expr(*value);
                } else {
                    self.constants.push(Literal::Void);
                    self.emit(Op::LOADCONST, &[self.constants.len() as u32 - 1]);
                }
                self.emit(Op::RET, &[]);
            }
            Expr::Label(label) => {
                let addr = self.code.len() as u32;
                self.labels.insert(
                    label.name.clone(),
                    Label {
                        high: (addr >> 8) & 0xff,
                        low: (addr & 0xff),
                    },
                );
            }
            Expr::Goto(goto) => {
                let label = self.labels.get(&goto.label);
                match label {
                    Some(l) => {
                        self.emit(Op::JUMP, &[l.high, l.low]);
                    }
                    None => {
                        panic!("Compiler: Label {} not found", goto.label);
                    }
                }
            }
            Expr::Extern(_) => {
                panic!("Only supported in Gos/Native");
            }
        }
    }

    fn patch_jump_addr(&mut self, pos: u32, addr: u32) -> () {
        self.code[pos as usize] = ((addr >> 8) & 0xff) as u8;
        self.code[(pos + 1) as usize] = (addr & 0xff) as u8;
    }
}

impl Bytecode {
    pub fn print(&self) -> () {
        println!("=== Constants ===");
        for (i, constant) in self.chunk.constants.iter().enumerate() {
            println!("  [{}] {:?}", i, constant);
        }
        println!("\n=== Bytecode ===");
        println!("Max Slot: {}", self.max_slot);

        let mut i = 0;
        while i < self.chunk.code.len() {
            let op_byte = self.chunk.code[i];
            if let Some(opcode) = Op::from_u8(op_byte) {
                let args_count = opcode.operand_count();
                print!("{:04x}: {:12}", i, Op::to_str(opcode.clone()));

                match opcode {
                    Op::LOADCONST => {
                        if i + 1 < self.chunk.code.len() {
                            let const_index = self.chunk.code[i + 1] as usize;
                            if const_index < self.chunk.constants.len() {
                                print!(
                                    " [{}] ; {:?}",
                                    const_index, self.chunk.constants[const_index]
                                );
                            } else {
                                print!(" [{}] ; INVALID", const_index);
                            }
                        }
                    }
                    Op::LOADVAR | Op::STOREVAR | Op::IN => {
                        if i + 1 < self.chunk.code.len() {
                            let slot = self.chunk.code[i + 1];
                            print!(" [slot {}]", slot);
                        }
                    }
                    Op::JUMP | Op::JUMPIFFALSE => {
                        if i + 2 < self.chunk.code.len() {
                            let addr = ((self.chunk.code[i + 1] as u16) << 8)
                                | (self.chunk.code[i + 2] as u16);
                            print!(" [addr {:04x}]", addr);
                        }
                    }
                    Op::CALL => {
                        if i + 3 < self.chunk.code.len() {
                            let addr = ((self.chunk.code[i + 1] as u16) << 8)
                                | (self.chunk.code[i + 2] as u16);
                            let arg_count = self.chunk.code[i + 3];
                            print!(" [addr {:04x}, {} args]", addr, arg_count);
                        }
                    }
                    _ => {
                        for j in 1..=args_count {
                            if i + j < self.chunk.code.len() {
                                print!(" {:02x}", self.chunk.code[i + j]);
                            }
                        }
                    }
                }
                println!();

                i += 1 + args_count;
            } else {
                println!("{:04x}: [UNKNOWN: {:02x}]", i, op_byte);
                i += 1;
            }
        }
    }
}
